### PR TITLE
Function references merge

### PR DESCRIPTION
**Pull request by dhil**
This PR merges `master` with the latest changes of `function-references/main`.

### CODE REVIEW

**Comment by dhil** [comment]
It seems the regression is related to imports of tags which carry payloads. The following program triggers the bug:
```wast
(module
  (tag $yield (export "yield"))
  (tag $spawn (export "spawn") (param i32))
)

(register "test")

(module
  (tag $spawn (import "test" "spawn") (param i32))
  (tag $yield (import "test" "yield")) ;; commenting this out makes the program validate.

  (elem declare func $main $thread1)

  (func $main
    (suspend $spawn (i32.const 0))
  )

  (func $thread1)
)
```

File: interpreter/valid/match.ml [hunk 1]
```diff
@@ -1,176 +1,183 @@
...
+  eq_limits c lim1 lim2
 
-and eq_global_type c a (GlobalType (t1, mut1)) (GlobalType (t2, mut2)) =
-  eq_mutability c a mut1 mut2 && eq_value_type c a t1 t2
+let eq_global_type c (GlobalT (mut1, t1)) (GlobalT (mut2, t2)) =
+  eq_mutability c mut1 mut2 && eq_val_type c t1 t2
 
-and eq_tag_type c a (TagType x1) (TagType x2) =
-  eq_var_type c a x1 x2
+and eq_tag_type c (TagT x1) (TagT x2) =
```
Comment (rossberg) [inline interpreter/valid/match.ml]:
```suggestion
let eq_tag_type c (TagT x1) (TagT x2) =
```

File: interpreter/valid/match.ml [hunk 2]
```diff
@@ -1,176 +1,183 @@
...
   match mut1 with
-  | Immutable -> match_value_type c a t1 t2
-  | Mutable -> eq_value_type c [] t1 t2
+  | Cons -> match_val_type c t1 t2
+  | Var -> eq_val_type c t1 t2
 
-and match_tag_type c a (TagType x1) (TagType x2) =
-  match_var_type c a x1 x2
+and match_tag_type c (TagT x1) (TagT x2) =
+  match_var_type c x1 x2
```
Comment (rossberg) [inline interpreter/valid/match.ml]:
```suggestion
let match_tag_type c tt1 tt2 =
  eq_tag_type c tt1 tt2
```

File: interpreter/valid/valid.ml [hunk 3]
```diff
@@ -820,27 +909,30 @@ let check_tag (c : context) (tag : tag) =
...
     check_memory_type c mt idesc.at;
-    {c with memories = mt :: c.memories}
+    {c with memories = c.memories @ [mt]}
   | GlobalImport gt ->
     check_global_type c gt idesc.at;
-    {c with globals = gt :: c.globals}
+    {c with globals = c.globals @ [gt]}
   | TagImport et ->
     check_tag_type c et idesc.at;
     {c with tags = et :: c.tags}
```
Comment (rossberg) [inline interpreter/valid/valid.ml]:
```suggestion
    {c with tags = c.tags @ [et]}
```

**Review by rossberg — APPROVED** [review]
Last comment fixes it. :)

**Comment by rossberg** [comment]
Thanks for grinding through this!

File: interpreter/valid/valid.ml (hunk 3 above)
Comment (dhil) [reply interpreter/valid/valid.ml]: Ah yikes! Thanks!

**Pull request merged**